File: noita-simulation/classes/particle.py
class Particle:
    def __init__(self, i, j) -> None:
        self.i = i
        self.j = j
        self.heat = 0
        self.lifespan = 0

    def moveToEmptyCell(self, targetI: int, targetJ: int, tiles: list[list]) -> list[list]:
        tiles[targetI][targetJ] = self #set self at target position
        tiles[self.i][self.j] = 0 #set current tile = 0
        self.i, self.j = targetI, targetJ #set self coords to target coords
        return tiles
    
    def switchWithAnotherParticle(self, targetI: int, targetJ: int, tiles: list[list]) -> list[list]:
        target = tiles[targetI][targetJ] #get target instance
        tiles[self.i][self.j] = target #set current tile = target
        tiles[targetI][targetJ] = self #set self at target position
        target.i, target.j = self.i, self.j #set target coords to current coords
        self.i, self.j = targetI, targetJ #set self coords to target coords
        return tiles
    
    def applyHeatToNeighborsIfIgnited(self, tiles : list[list]):
        pass

    def checkLifeSpan(self): 
        pass

    def takeDamage(self):
        pass

    def update(self):
        pass
    
    def getNeighbors(self):
        pass
    

File: noita-simulation-revamped/classes/Particle.java
package classes;

public class Particle {
    public int i;
    public int j;

    //equivalent to def __init__
    public Particle(int i, int j){
        this.i = i;
        this.j = j;
    }

    public Object[][] moveToEmptyCell(int targetI, int targetJ, Object[][] tiles){
        tiles[targetI][targetJ] = this; // set self at target position
        tiles[this.i][this.j] = 0; // set current tile = 0
        this.i = targetI; //set self coords to target coords
        this.j = targetJ;
        
        return tiles;
    }

    public Object[][] switchWithAnotherParticle(int targetI, int targetJ, Object[][] tiles) {
        Particle target = (Particle) tiles[targetI][targetJ]; //get target instance
        tiles[this.i][this.j] = target; //set current tile = target
        tiles[targetI][targetJ] = this; //set self at target position
        
        target.i = this.i; //set target coords to current coords
        target.j = this.j; 

        this.i = targetI; // set self coords to target coords
        this.j = targetJ;

        return tiles;
    }

    

    
}
